src/mds/CInode.h

```diff
@@ -385,108 +388,104 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
-  int num_projected_xattrs;
-  int num_projected_srnodes;
-  
-  inode_t *project_inode(std::map<std::string,bufferptr> *px=0);
+
+private:
+  mempool::mds_co::list<projected_inode> projected_nodes;   // projected values (only defined while dirty)
+  size_t num_projected_xattrs = 0;
+  size_t num_projected_srnodes = 0;
+
+  sr_t &project_snaprealm(projected_inode &pi);
+public:
+  CInode::projected_inode &project_inode(bool xattr = false, bool snap = false);
   void pop_and_dirty_projected_inode(LogSegment *ls);
 
-  projected_inode_t *get_projected_node() {
+  projected_inode *get_projected_node() {
     if (projected_nodes.empty())
       return NULL;
     else
-      return projected_nodes.back();
+      return &projected_nodes.back();
   }
 
   version_t get_projected_version() const {
     if (projected_nodes.empty())
       return inode.version;
     else
-      return projected_nodes.back()->inode->version;
+      return projected_nodes.back().inode.version;
   }
   bool is_projected() const {
     return !projected_nodes.empty();
   }
 
-  const inode_t *get_projected_inode() const {
+  const mempool_inode *get_projected_inode() const {
     if (projected_nodes.empty())
       return &inode;
     else
-      return projected_nodes.back()->inode;
+      return &projected_nodes.back().inode;
   }
-  inode_t *get_projected_inode() {
+  mempool_inode *get_projected_inode() {
     if (projected_nodes.empty())
       return &inode;
     else
-      return projected_nodes.back()->inode;
+      return &projected_nodes.back().inode;
   }
-  inode_t *get_previous_projected_inode() {
+  mempool_inode *get_previous_projected_inode() {
     assert(!projected_nodes.empty());
-    std::list<projected_inode_t*>::reverse_iterator p = projected_nodes.rbegin();
-    ++p;
-    if (p != projected_nodes.rend())
-      return (*p)->inode;
+    auto it = projected_nodes.rbegin();
+    ++it;
+    if (it != projected_nodes.rend())
+      return &it->inode;
     else
       return &inode;
   }
 
-  std::map<std::string,bufferptr> *get_projected_xattrs() {
+  mempool_xattr_map *get_projected_xattrs() {
     if (num_projected_xattrs > 0) {
-      for (std::list<projected_inode_t*>::reverse_iterator p = projected_nodes.rbegin();
-	   p != projected_nodes.rend();
-	   ++p)
-	if ((*p)->xattrs)
-	  return (*p)->xattrs;
+      for (auto it = projected_nodes.rbegin(); it != projected_nodes.rend(); ++it)
+	if (it->xattrs)
+	  return it->xattrs.get();
     }
     return &xattrs;
   }
-  std::map<std::string,bufferptr> *get_previous_projected_xattrs() {
-      std::list<projected_inode_t*>::reverse_iterator p = projected_nodes.rbegin();
-    for (++p;  // skip the most recent projected value
-	 p != projected_nodes.rend();
-	 ++p)
-      if ((*p)->xattrs)
-	return (*p)->xattrs;
+  mempool_xattr_map *get_previous_projected_xattrs() {
+    if (num_projected_xattrs > 0) {
+      for (auto it = ++projected_nodes.rbegin(); it != projected_nodes.rend(); ++it)
```

Comment:
why ++projected_nodes.rbegin()

Reply:
Original code `get_previous_projected_xattrs` increments the iterator at the beginning (hence "previous"). I just moved it to the initialization of the for loop because it's more compact.

Reply:
sorry, I mean the '++projected_nodes.rbegin()' in get_projected_srnode()